src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ApplicationPropertyPage/CompoundTargetFrameworkValueProvider.cs

```diff
@@ -0,0 +1,152 @@
+            ComplexTargetFramework values = BreakDownValue(storedComplexValue);
+            return await base.OnGetUnevaluatedPropertyValueAsync(propertyName, ExtractValue(values, propertyName) ?? "", defaultProperties);
+        }
+
+        //private static string? CommonCode(string propertyName, IProjectProperties defaultProperties)
+        //{
+        //    string? storedComplexValue = defaultProperties.GetUnevaluatedPropertyValueAsync(PropertyToBeSavedOnProjectFile); // TODO: change to TargetFrameworkProperty
+        //    ComplexTargetFramework values = BreakDownValue(storedComplexValue);
+        //    return ExtractValue(values, propertyName);
+        //}
+
+        /// <summary>
+        /// Extracts the specified value from the ComplexTargetFramework structure.
+        /// </summary>
+        /// <param name="values">The strucutre were all the values are.</param>
+        /// <param name="propertyName">The property's value we want to return.</param>
+        /// <returns></returns>
+        private static string? ExtractValue(ComplexTargetFramework values, string propertyName)
+        {
+            return propertyName switch
+            {
+                TargetFrameworkProperty => values.TargetFramework,
+                TargetOSProperty => values.Platform,
+                TargetOSVersionProperty => values.PlatformVersion,
+                _ => null,
+            };
+        }
+
+        /// <summary>
+        /// Combines the three values that make up the TargetFramework property.
+        /// I.e. net5.0, windows, 10.0 => net5.0-windows10.0
+        /// net5.0, null, null => net5.0
+        /// </summary>
+        /// <param name="targetFramework">The target framework value.</param>
+        /// <param name="targetPlatform">The target platform value.</param>
+        /// <param name="platformVersion">The platform version value.</param>
+        /// <returns></returns>
+        private static string ComputeValue(string targetFramework, string? targetPlatform, string? platformVersion)
+        {
+            if (targetPlatform != null) // TODO: Ensure TargetFramework is not null.
+            {
+                return targetFramework + "-" + targetPlatform + platformVersion;
+            }
+            return targetFramework;
+        }
+
+        /// <summary>
+        /// Takes the combined string and separates the values into the ComplextTargetFramework structure.
+        /// TargetFramework must always be present, but Platform and PlatformVersion can be null.
+        /// <para>i.e. net5.0-windows10.0 => { net5.0, windows, 10.0 }; netcoreapp3.1 => { netcoreapp3.1, null, null }</para>
+        /// </summary>
+        /// <param name="storedComplexTargetFramework"></param>
+        /// <returns></returns>
+        private static ComplexTargetFramework BreakDownValue(string? storedComplexTargetFramework)
+        {
+            ComplexTargetFramework result = new ComplexTargetFramework();
```

Comment:
This string is a user property and should not be parsed; https://github.com/dotnet/project-system/blob/main/docs/repo/coding-conventions.md#data. Read the associated properties instead.

Reply:
`TargetFramework` can contain `FooBar` and as long as the those previous properties are set, everything should just work.

Reply:
This was _super_ valuable, and it's addressed in the new commits, thanks for the feedback!